File: java/minimum_window_substring.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;

public class HelloWorld {
    public String minWindow(String s, String t) {
        Map<Character, Integer> count=new HashMap<Character, Integer>();
        for (char c:t.toCharArray()) {
            Integer cnt=count.get(c);
            if (cnt==null) {
                count.put(c, 1);
            } else {
                count.put(c, cnt+1);
            }
        }
        
        int total=0;        
        int start=0, end=0, minlen=Integer.MAX_VALUE;
        for (int l=0, r=0; r<s.length(); r++) {
            if (count.containsKey(s.charAt(r))) {
                count.put(s.charAt(r), count.get(s.charAt(r))-1);
                if (count.get(s.charAt(r))>=0) total++;
            }
            
            while (total==t.length()) {
                if (minlen>r-l+1) {
                    minlen=r-l+1;
                    start=l;
                    end=r;
                }
                
                if (count.containsKey(s.charAt(l))) {
                    count.put(s.charAt(l), count.get(s.charAt(l))+1);
                    if (count.get(s.charAt(l))>0) total--;
                }
                
                l++;
            }
        }
        
        if (minlen==Integer.MAX_VALUE) return "";
        else return s.substring(start, end+1);
    }
    
    public static void main(String[] args) {
        HelloWorld h=new HelloWorld();
        String res=h.minWindow("a", "a");
    }
}

File: python/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        l, total, minLen, start=0, len(t), float('inf'), 0
        d=collections.Counter(t)
        
        for r in range(len(s)):
            if s[r] in d:
                d[s[r]]-=1
                if d[s[r]]>=0: total-=1
            if total==0:
                while total==0:
                    if s[l] in d:
                        d[s[l]]+=1
                        if d[s[l]]>0: total+=1
                    l+=1
                if minLen>r-l+2:
                    minLen=r-l+2
                    start=l-1
        return s[start:start+minLen] if minLen!=float('inf') else ""
    
